src/sleep/sleep.c

```
@@ -38,8 +39,192 @@
+                if (fiemap->fm_mapped_extents) {
+                        const __u32 i = fiemap->fm_mapped_extents - 1;
+
+                        fiemap_start = fiemap->fm_extents[i].fe_logical +
+                                       fiemap->fm_extents[i].fe_length;
+
+                        if (fiemap->fm_extents[i].fe_flags & FIEMAP_EXTENT_LAST)
+                                break;
+                }
+        }
+
+        result_fiemap->fm_mapped_extents = result_extents;
+        free(fiemap);
+
+        return result_fiemap;
+
+fail_cleanup:
+        if (result_fiemap)
+                free(result_fiemap);
+
+        if (fiemap)
+                free(fiemap);
+
+        return NULL;
+}
+
+static int write_disk_offset(void) {
+        _cleanup_free_ char *device = NULL, *type = NULL, *write = NULL;
+        _cleanup_free_ struct fiemap *fiemap = NULL;
+        struct stat stb;
+        uint64_t offset;
+        int fd;
+        int r;
+
+        /* Only available in 4.17+ */
+        r = access("/sys/power/disk_offset", W_OK);
+        if (r < 0) {
+                log_debug_errno(r, "/sys/power/disk_offset unavailable: %m");
+                return 0;
+        }
+        r = find_hibernate_location(&device, &type, NULL, NULL);
+        if (r < 0)
+                return 0;
+
+        /* if it's a swap partition, we don't append offset */
+        if (strcmp(type, "partition") == 0)
+                return write_string_file("/sys/power/disk_offset", device, 0);
+        else if(strcmp(type, "file") != 0) {
+                log_error_errno(r, "Unrecognized swap type '%s' %m", type);
+                return -EINVAL;
+        }
+
+        r = stat(device, &stb);
+        if (r < 0) {
+                log_debug("Unable to stat %s", device);
+                return r;
+        }
+        fd = open(device, O_RDONLY);
+        if (fd < 0) {
+                log_debug_errno(fd, "Unable to open '%s' %m", device);
+                return -EBADF;
```

please propagate the right error here, i.e. "return -errno".